src/factorizations/blocklanczos.jl

```diff
@@ -0,0 +1,326 @@
+
+Structure for storing vectors in a block format. The type parameter `T` represents the type of vector elements,
+while `S` represents the type of inner products between vectors.
+"""
+struct BlockVec{T,S<:Number}
+    vec::Vector{T}
+    function BlockVec{S}(vec::Vector{T}) where {T,S<:Number}
+        return new{T,S}(vec)
+    end
+end
+Base.length(b::BlockVec) = length(b.vec)
+Base.getindex(b::BlockVec, i::Int) = b.vec[i]
+function Base.getindex(b::BlockVec{T,S}, idxs::AbstractVector{Int}) where {T,S}
+    return BlockVec{S}([b.vec[i] for i in idxs])
+end
+Base.setindex!(b::BlockVec{T}, v::T, i::Int) where {T} = (b.vec[i] = v)
+function Base.setindex!(b₁::BlockVec{T}, b₂::BlockVec{T},
+                        idxs::AbstractVector{Int}) where {T}
+    return (b₁.vec[idxs] = b₂.vec;
+            b₁)
+end
+LinearAlgebra.norm(b::BlockVec) = norm(b.vec)
+function apply(f, block::BlockVec{T,S}) where {T,S}
+    return BlockVec{S}([apply(f, block[i]) for i in 1:length(block)])
+end
+function initialize(x₀, size::Int)
+    S = typeof(inner(x₀, x₀))
+    x₀_vec = [randn!(similar(x₀)) for _ in 1:(size - 1)]
+    pushfirst!(x₀_vec, x₀)
+    return BlockVec{S}(x₀_vec)
+end
+function Base.push!(V::OrthonormalBasis{T}, b::BlockVec{T}) where {T}
+    for i in 1:length(b)
+        push!(V, b[i])
+    end
+    return V
+end
+function Base.copy(b::BlockVec)
+    return BlockVec{typeof(b).parameters[2]}(scale.(b.vec, 1))
+end
+Base.iterate(b::BlockVec) = iterate(b.vec)
+Base.iterate(b::BlockVec, state) = iterate(b.vec, state)
+
+"""
+    mutable struct BlockLanczosFactorization{T,S<:Number,SR<:Real} <: BlockKrylovFactorization{T,S,SR}
+
+Structure to store a BlockLanczos factorization of a real symmetric or complex hermitian linear
+map `A` of the form
+
+```julia
+A * V = V * B + r * b'
+```
+
+For a given BlockLanczos factorization `fact`, length `k = length(fact)` and basis `V = basis(fact)` are
+like [`LanczosFactorization`](@ref). The block tridiagonal matrix `TDB` is preallocated in `BlockLanczosFactorization`
+and is of type `Hermitian{S<:Number}` with `size(TDB) == (k,k)`. The residuals `r` is of type `Vector{T}`.
+One can also query [`normres(fact)`](@ref) to obtain `norm(r)`, the norm of the residual. The matrix
+`b` takes the default value ``[0;I]``, i.e. the matrix of size `(k,bs)` and an unit matrix in the last
+`bs` rows and all zeros in the other rows. `bs` is the size of the last block. One can query [`r_size(fact)`] to obtain
+size of the last block and the residuals.
+
+`BlockLanczosFactorization` is mutable because it can [`expand!`](@ref). But it does not support `shrink!`
+because it is implemented in its `eigsolve`.
+See also [`BlockLanczosIterator`](@ref) for an iterator that constructs a progressively expanding
+BlockLanczos factorizations of a given linear map and a starting vector.
+"""
+mutable struct BlockLanczosFactorization{T,S<:Number,SR<:Real} <:
+               KrylovFactorization{T,S}
+    total_size::Int
```

Comment:
Can we keep the name `k` for this one as for `Lanczos`? I should probably have a more descriptive name but I would then modify it everywhere across Lanczos, Arnoldi and BlockLanczos. 

```suggestion
    k::Int
```